relay/src/main.rs

```diff
@@ -98,99 +147,146 @@ async fn main() -> Result<(), Error> {
-
-            // upload binary to Bonsai
-            let bonsai_client = get_client_from_parts(bonsai_api_url, bonsai_api_key).await?;
-            let img_id = image_id.clone();
-            match put_image(
-                bonsai_client.clone(),
-                img_id.clone(),
-                guest_entry.elf.to_vec(),
+        Command::Upload { guest_binary } => {
+            let image_ids = upload_images(
+                guest_binary,
+                &args.global_opts.bonsai_api_url,
+                &args.global_opts.bonsai_api_key,
             )
-            .await
-            {
-                Ok(()) => (),
-                Err(SdkErr::ImageIdExists) => (),
-                Err(err) => return Err(err.into()),
-            }
-
-            print!("{image_id}");
+            .await?;
+
+            let output = hex::encode(ethers::abi::encode(&[Token::Array(
+                image_ids
+                    .into_iter()
+                    .map(|image_id| {
+                        Hash::from(bytemuck::cast::<_, [u8; 32]>(image_id)).into_token()
+                    })
+                    .collect(),
+            )]));
+            print!("{output}");
             std::io::stdout()
                 .flush()
-                .context("Failed to flush stdout buffer")?;
+                .context("failed to flush stdout buffer")?;
         }
         Command::Run {
-            bonsai_api_url,
-            bonsai_api_key,
             relay_address,
             eth_node,
             eth_chain_id,
             private_key,
         } => {
-            let ethers_client = create_ethers_client_private_key(
-                &eth_node.unwrap_or_else(|| "ws://localhost:8545".to_string()),
-                &private_key.unwrap_or_else(|| {
-                    "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80".to_string()
-                }),
-                eth_chain_id,
-            )
-            .await?;
-
             let relayer = Relayer {
-                publish_mode: true,
-                publish_port: "8080".to_string(),
-                bonsai_api_url: bonsai_api_url
-                    .unwrap_or_else(|| "http://localhost:8081".to_string()),
-                bonsai_api_key: bonsai_api_key.unwrap_or_else(|| "".to_string()),
+                rest_api: true,
+                dev_mode: dev_mode,
+                rest_api_port: "8080".to_string(),
+                bonsai_api_url: args.global_opts.bonsai_api_url.clone(),
+                bonsai_api_key: args.global_opts.bonsai_api_key.clone(),
                 relay_contract_address: relay_address,
             };
-            let _ = tokio::spawn(relayer.run(ethers_client.clone())).await;
+            let client_config =
+                EthersClientConfig::new(eth_node, eth_chain_id, private_key.try_into()?);
+            let server_handle = tokio::spawn(relayer.run(client_config));
+
+            // HACK: Wait 1 second to give local Bonsai a chance to start.
```

Comment:
Is this for bonsai or the relay to startup?

Reply:
Technically it's for the thread managing the rest-API-mock of Bonsai inside the Relay

Reply:
ahhh cool!